ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
+  # Creates a new instance of a Client
+  def self.new(config: Temporal::Configuration, **options) -> Temporal::Client
+
+  # Start the workflow and return handle for further interaction
+  # TODO: expand on :options
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Temporal::WorkflowHandle
+
+  # Start the workflow and block awaiting it's completion
+  def execute_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Any
+
+  # Generate a handle object from a namespace, workflow_id and a run_id
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> Temporal::WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both positional arguments and keyword arguments are supported, except for the reserved `options:`
+- Conventions similar to Python SDK are used here:
+  - `Client` is a general purpose interface, not limited workflow-only interactions
+  - Due to this method names are more descriptive than just `#start`, `#execute`, etc
+  - `Client#execute_workflow` is added for convenience
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a workflow_id/run_id combination using the
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if a workflow is still running after
+  #   waiting for for up to :timeout seconds.
+  # Passing timeout: nil will not wait if the working is still running
+  def result(timeout: Integer) -> Any
```

Comment:
I'm not sure we want to expose `timeout` here, we could use grpc DEADLINE instead which is a well establish concept and applies to all client calls.

We should also add a `follow_runs: boolean` parameter if you don't want to wait for the entire execution chain to complete (e.g. with a cron workflow).

We don't have the option to not wait for the result at all in other SDKs (with this API) I'm wondering how useful it is and if we should add it to the other SDKs.

Reply:
`follow_runs` is implemented in the client and is the default behavior in other SDKs.
See TS implementation for reference: https://github.com/temporalio/sdk-typescript/blob/a2020cc88d6099443cf72443b03f4b9fe186f875/packages/client/src/workflow-client.ts#L398.

As for timeout as a param, it conflicts with the grpc deadline concept which serves the same purpose but is applicable to any grpc call, I'm suggesting we only use grpc deadline and drop this alternative way of setting a timeout.

Reply:
I'll adopt the `follow_runs`, sounds good 👍 

As for the gRPC deadline — are you suggesting using `deadline:` instead of `timeout:` argument? Or that a deadline should be set per gRPC client (for all the calls made)?

Reply:
@bergundy I disagree with setting deadline here actually - gRPC is an implementation detail of our platform (theoretically) and that we might turn a timeout (more generic term) into a gRPC deadline is incidental to the user.

It also better facilitates concepts like auto-retries of the kind Max talked about where we use different, smaller deadlines underneath the overall timeout.

Reply:
@Sushisource how is this call different then the other APIs?
In all of the other SDKs we respect grpc deadline, I don't see why we'd want to hide this.
We're actually taking a step in the opposite direction, e.g. exposing the grpc client and allowing setting grpc interceptors in golang.

Reply:
It isn't - I think they should all work that way _if you are using a higher-level client_. Exposing the raw client makes plenty of sense for when people need it, and of course that would operate in terms of deadlines because that's how gRPC clients work - but IMO our "friendly" client is ideally transport/protocol agnostic.

Get result happens to be a pretty good example of something that can really only exist on the "higher level" client since it needs to do all sorts of stuff

Reply:
I agree, this timeout is actually different from grpc DEADLINE because under the hood this API can make multiple requests, each with a separate DEADLINE, but... any grpc request executed in this context should respect the context's DEADLINE regardless of this timeout param, so I'm a bit on the fence whether we want to have 2 possibly conflicting ways to time out this operation.